patchwork-capabilities/src/main/java/net/minecraftforge/common/capabilities/CapabilityDispatcher.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Minecraft Forge, Patchwork Project
+ * Copyright (c) 2016-2019, 2019
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.capabilities;
+
+import java.util.List;
+import java.util.Map;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.ParametersAreNonnullByDefault;
+
+import com.google.common.collect.Lists;
+import net.minecraftforge.common.util.INBTSerializable;
+import net.minecraftforge.common.util.LazyOptional;
+
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.Tag;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.math.Direction;
+
+/**
+ * A high-speed implementation of a {@link Capability} delegator.
+ * This is used to wrap the results of the {@link net.minecraftforge.event.AttachCapabilitiesEvent}.
+ * It is HIGHLY recommended that you DO NOT use this approach unless
+ * you MUST delegate to multiple providers instead just implementing
+ * your handlers using normal if statements.
+ *
+ * <p>Internally the handlers are baked into arrays for fast iteration.
+ * The {@link Identifier} will be used for the NBT Key when serializing.
+ */
+@ParametersAreNonnullByDefault
+public final class CapabilityDispatcher implements INBTSerializable<CompoundTag>, ICapabilityProvider {
+	private final ICapabilityProvider[] caps;
+	private final INBTSerializable<Tag>[] writers;
+	private final String[] names;
+	private final List<Runnable> listeners;
+
+	public CapabilityDispatcher(Map<Identifier, ICapabilityProvider> providers, List<Runnable> listeners) {
+		this(providers, listeners, null);
+	}
+
+	@SuppressWarnings("unchecked")
+	public CapabilityDispatcher(Map<Identifier, ICapabilityProvider> providers, List<Runnable> listeners, @Nullable ICapabilityProvider parent) {
+		List<ICapabilityProvider> lstCaps = Lists.newArrayList();
```

Comment:
Can we drop the `lstX` prefix here, and then just use `this.X` at the end instead?